File: app/src/FizzBuzzCommand.ts
import { FizzBuzzValue } from './FizzBuzzValue'
import { FizzBuzzList } from './FizzBuzzList'
import { FizzBuzzType } from './FizzBuzzType'

export interface FizzBuzzCommand {
  execute(): FizzBuzzValue | FizzBuzzList
}

export class FizzBuzzValueCommand implements FizzBuzzCommand {
  private fizzBuzzType: FizzBuzzType
  private number: number

  constructor(fizzBuzzType: FizzBuzzType, number: number) {
    this.fizzBuzzType = fizzBuzzType
    this.number = number
  }

  execute(): FizzBuzzValue {
    return this.fizzBuzzType.generate(this.number)
  }
}

export class FizzBuzzListCommand implements FizzBuzzCommand {
  private fizzBuzzType: FizzBuzzType
  private start: number
  private end: number

  constructor(fizzBuzzType: FizzBuzzType, start: number = 1, end: number = 100) {
    this.fizzBuzzType = fizzBuzzType
    this.start = start
    this.end = end
  }

  execute(): FizzBuzzList {
    let list = new FizzBuzzList()
    for (let i = this.start; i <= this.end; i++) {
      const value = this.fizzBuzzType.generate(i)
      list = list.add(value)
    }
    return list
  }
}

File: app/kotlin-fizzbuzz/src/main/kotlin/application/FizzBuzzCommand.kt
package application

import domain.model.FizzBuzzList
import domain.model.FizzBuzzValue
import domain.type.FizzBuzzType

class AssertionFailedException(message: String) : Exception(message)

object Assertions {
    fun require(condition: Boolean, message: String = "Assertion failed") {
        if (!condition) {
            throw AssertionFailedException(message)
        }
    }
}

interface FizzBuzzCommand {
    fun execute(): Any
}

class FizzBuzzValueCommand(private val type: FizzBuzzType) : FizzBuzzCommand {
    override fun execute(): Any {
        throw UnsupportedOperationException("Use execute(number) instead")
    }
    
    fun execute(number: Int): FizzBuzzValue {
        Assertions.require(number > 0, "値は正の値のみ許可する")
        val value = type.generate(number)
        return FizzBuzzValue(number, value)
    }
}

class FizzBuzzListCommand(private val type: FizzBuzzType) : FizzBuzzCommand {
    companion object {
        const val DEFAULT_END_NUMBER = 100
    }
    
    override fun execute(): FizzBuzzList {
        val values = (1..DEFAULT_END_NUMBER).map { number ->
            FizzBuzzValueCommand(type).execute(number).value
        }
        return FizzBuzzList(values)
    }
}